Optimizes the performance of circular buffer

iansseijelly opened the pull request:
# Description

This PR addresses issue #4274.


## Type of change
- Bug fix (non-breaking change which fixes an issue)

## Screenshots

Training throughput before and after the patch running training on task `[Isaac-Velocity-Flat-Spot-v0]`.

<img width="1360" height="802" alt="delayed_buffer_perf_fix" src="https://github.com/user-attachments/assets/34e2d313-1317-48e4-8278-8729c8a7315b" />
## Checklist

- [x] I have read and understood the [contribution guidelines](https://isaac-sim.github.io/IsaacLab/main/source/refs/contributing.html)
- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [x] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

iansseijelly committed:
perf: optimize the performance of circular buffer to cut its impact in the hot path, avoiding unnecessary kernel and synchronization

iansseijelly committed:
fix: run pre commit and change contributors

garylvov commented:
Hi, this is a cool optimization, I'm excited to try it, thank you!

Two quick qs: 

- Does this only affect configurations that use the delayed PD actuator, or does it apply to all actuator configurations?
- I see in your charts that the FPS is improved. How much of a difference does this make in cumulative training time? Like, does it go from 2 hours to train to 1.5 hours to train? I'd love to see the reward with and without the optimization over time

Mayankm96 commented on source/isaaclab/isaaclab/utils/buffers/circular_buffer.py:
```diff
@@ -100,6 +104,8 @@ def reset(self, batch_ids: Sequence[int] | None = None):
         # reset the number of pushes for the specified batch indices
         self._num_pushes[batch_ids] = 0
+        # re-initialization is required
+        self._all_initialized = False
```
NIT: Instead of "all_initialized", a better name might be "_has_reset"?

Syntax-wise, initialization sounds more like creation of the buffers, while here it is more resetting them.

T-K-233 commented on source/isaaclab/isaaclab/actuators/actuator_pd.py:
```diff
@@ -358,9 +358,9 @@ def compute(
-        control_action.joint_positions = self.positions_delay_buffer.compute(control_action.joint_positions)
-        control_action.joint_velocities = self.velocities_delay_buffer.compute(control_action.joint_velocities)
-        control_action.joint_efforts = self.efforts_delay_buffer.compute(control_action.joint_efforts)
+        control_action.joint_positions[:] = self.positions_delay_buffer.compute(control_action.joint_positions)
```
This uses slice assignment to keep the original tensor storage (`control_action.joint_positions` etc.) and overwrite its contents in place. Without this, compute() may return a new tensor, causing buffer replacement and additional allocation or copy overhead.

Mayankm96 replied:
@iansseijelly and @T-K-233 : Let me know your thoughts on the above. Thanks again!

iansseijelly committed:
fix: change the naming from _all_initialized to _need_reset, hence inverting all boolean conditions

Mayankm96 commented on source/isaaclab/isaaclab/utils/buffers/circular_buffer.py:
```diff
@@ -100,6 +104,8 @@ def reset(self, batch_ids: Sequence[int] | None = None):
             batch_ids = slice(None)
         # reset the number of pushes for the specified batch indices
         self._num_pushes[batch_ids] = 0
+        # re-initialization is required
```
```suggestion
        # reset is needed on next update to fill entire buffer with initial data
```

iansseijelly committed:
fix: changing comments from init to reset vacb

iansseijelly commented on source/isaaclab/isaaclab/utils/buffers/circular_buffer.py:
```diff
@@ -129,9 +135,14 @@ def append(self, data: torch.Tensor):
+            if is_first_push.any().item():
+                self._buffer[:, is_first_push] = data[is_first_push]
+            else:
+                # All batches now initialized, skip this check in future calls
```
Thanks, I have rephrased all the comments accordingly.

AntoineRichard reviewed (commented):
The policies are training well on this branch and they do transfer to main, so it's likely the implementation is correct.

Code-wise it also looks good. I've suggested some modification to further improve the performance, but it's not as substantial as removing the cloning step. If we are in a rush we can get this in as is. These will likely get reworked to support warp anyway.

iansseijelly commented on source/isaaclab/isaaclab/utils/buffers/circular_buffer.py:
```diff
@@ -100,6 +104,8 @@ def reset(self, batch_ids: Sequence[int] | None = None):
+        self._need_reset = True
         if self._buffer is not None:
             # set buffer at batch_id reset indices to 0.0 so that the buffer() getter returns the cleared circular buffer after reset.
             self._buffer[:, batch_ids, :] = 0.0
```
I agree with this. Thanks for the detailed microbenchmark.

iansseijelly commented on source/isaaclab/isaaclab/utils/buffers/circular_buffer.py:
```diff
@@ -129,9 +135,14 @@ def append(self, data: torch.Tensor):
+        if self._need_reset:
+            is_first_push = self._num_pushes == 0
+            if is_first_push.any().item():
+                self._buffer[:, is_first_push] = data[is_first_push]
```
I also agree with this. Looks like it saves a lot of time. 
Probably need a little more comment/documentation on what's happening here since mask programming is less interpretable.

T-K-233 committed:
add: address performance optimization comments suggested

iansseijelly commented:
Hi, we have just addressed the suggestions that @AntoineRichard provided. 
Thanks for the help, everyone :)

iansseijelly commented:
> @iansseijelly Thanks for the improvements and addressing the comments! As we've shifted our development to the new `develop` branch, could you do a rebase to resolve the conflicts on top of develop?

We just rebased atop `develop`. Thanks.

kellyguo11 commented on source/isaaclab/isaaclab/utils/buffers/circular_buffer.py:
```diff
@@ -156,8 +167,9 @@ def __getitem__(self, key: torch.Tensor) -> torch.Tensor:
-            raise RuntimeError("Attempting to retrieve data on an empty circular buffer. Please append data first.")
+        if self._need_reset:
+            if self._buffer is None or (self._num_pushes == 0).any().item():
+                raise RuntimeError("Attempting to retrieve data on an empty circular buffer. Please append data first.")
```
could we simplify this to check only for `self._need_reset` then if that's the only value needed?

T-K-233 committed:
add: remove the in-place copy, since compute slicing time_lags already returns a copy.

T-K-233 committed:
fix: remove unnecessary checks

kellyguo11 commented:
hmm github is not giving me the option to update the branch. could you grab the latest changes from tot develop so that the CI tests can run? we added a fix recently that was preventing the installation from working.

kellyguo11 commented:
Thanks a lot! seeing a test failure in test_observation_manager.py

```
source/isaaclab/test/managers/test_observation_manager.py ......F....... [ 93%]
.                                                                        [100%]

=================================== FAILURES ===================================
_________________________ test_compute_with_2d_history _________________________
source/isaaclab/test/managers/test_observation_manager.py:462: in test_compute_with_2d_history
    observations = obs_man.compute()
                   ^^^^^^^^^^^^^^^^^
source/isaaclab/isaaclab/managers/observation_manager.py:337: in compute
    obs_buffer[group_name] = self.compute_group(group_name, update_history=update_history)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
source/isaaclab/isaaclab/managers/observation_manager.py:421: in compute_group
    circular_buffer.append(obs)
source/isaaclab/isaaclab/utils/buffers/circular_buffer.py:145: in append
    expanded_mask = is_first_push[None, :, None].expand_as(self._buffer)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   RuntimeError: The expanded size of the tensor (256) must match the existing size (20) at non-singleton dimension 3.  Target sizes: [5, 20, 128, 256, 1].  Tensor sizes: [1, 20, 1]
```

Dexsuite task in test_environments.py also failing with similar error:

```
________ test_environments[Isaac-Dexsuite-Kuka-Allegro-Lift-v0-32-cuda] ________
source/isaaclab_tasks/test/test_environments.py:38: in test_environments
    _run_environments(task_name, device, num_envs, create_stage_in_memory=False)
source/isaaclab_tasks/test/env_test_utils.py:158: in _run_environments
    _check_random_actions(
source/isaaclab_tasks/test/env_test_utils.py:230: in _check_random_actions
    obs, _ = env.reset()
             ^^^^^^^^^^^
/isaac-sim/kit/python/lib/python3.12/site-packages/gymnasium/wrappers/common.py:400: in reset
    return super().reset(seed=seed, options=options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
/isaac-sim/kit/python/lib/python3.12/site-packages/gymnasium/core.py:333: in reset
    return self.env.reset(seed=seed, options=options)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
source/isaaclab/isaaclab/envs/manager_based_env.py:384: in reset
    self.obs_buf = self.observation_manager.compute(update_history=True)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
source/isaaclab/isaaclab/managers/observation_manager.py:337: in compute
    obs_buffer[group_name] = self.compute_group(group_name, update_history=update_history)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
source/isaaclab/isaaclab/managers/observation_manager.py:412: in compute_group
    circular_buffer.append(obs)
source/isaaclab/isaaclab/utils/buffers/circular_buffer.py:145: in append
    expanded_mask = is_first_push[None, :, None].expand_as(self._buffer)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
E   RuntimeError: The expanded size of the tensor (4) must match the existing size (32) at non-singleton dimension 2.  Target sizes: [5, 32, 4, 3].  Tensor sizes: [1, 32, 1]
```

T-K-233 committed:
fix: revert logic to correctly handle higher dimensional tensor

StafaH commented:
Hi @iansseijelly, sorry for the delay in getting back to you here. We are still interested in getting this merged in, are you able to update the MR and resolve the conflicts? If you can also post up-to-date benchmark results in the perf difference that would be appreciated!

ooctipus commented:
Thanks @iansseijelly (and apologies for the long open time) — the ideas here have since landed on `develop` via the frame-stacking refactor (#5574, #5849): `_max_len_int` avoids the `.item()` sync in `max_length`, and the CPU-side `_need_reset` gate skips the steady-state `torch.any` probe. The final piece this PR targeted — the `torch.any` + masked backfill inside `append` — is removed in #6471 with benchmarks. Closing as superseded; the analysis here was correct and is now fully covered on `develop`.
